[STORM-2172][SQL] Support Avro as input / output format

vesense opened the pull request:
- new AvroScheme & AvroSerializer , unit tests
- new SerdeUtils

vesense commented:
After https://github.com/apache/storm/pull/1748 getting merged, I will update this PR accordingly.

sachin-goyal commented on external/sql/storm-sql-runtime/src/jvm/org/apache/storm/sql/runtime/serde/avro/AvroScheme.java:
```diff
@@ -0,0 +1,90 @@
+            Object key = entry.getKey();
+            Object newKey = key instanceof Utf8 ? key.toString() : key;
+            Object val = entry.getValue();
+            Object newVal = val instanceof Utf8 ? val.toString() : val;
```
If we are converting key and val to String from Utf8, then do we need to recursively check the existence of all Utf8 type values inside key/val objects?

vesense replied:
Good catch. Updated.

sachingsachin commented on external/sql/storm-sql-runtime/src/jvm/org/apache/storm/sql/runtime/serde/avro/AvroScheme.java:
```diff
@@ -0,0 +1,81 @@
+          Map<Object, Object> map = SerdeUtils.convertAvroUtf8Map((Map<Object, Object>)value);
+          list.add(map);
+        } else {
+          list.add(value);
```
Do we need to check array type containers too for Utf8-to-String conversion? Example: array or set of Utf-8 objects? That should cover all the cases IMO

sachingsachin commented on external/sql/storm-sql-runtime/src/jvm/org/apache/storm/sql/runtime/utils/SerdeUtils.java:
```diff
@@ -0,0 +1,100 @@
+        for (Map.Entry<Object, Object> entry : value.entrySet()) {
+            // Avro only allows maps with Strings for keys, so we only have to worry
+            // about deserializing the values
+            Object key = entry.getKey().toString();
```
Avro supports non-string maps too. So conversion to string might be incorrect in those cases.

vesense replied:
Updated.

sachingsachin commented on external/sql/storm-sql-runtime/src/jvm/org/apache/storm/sql/runtime/utils/SerdeUtils.java:
```diff
@@ -0,0 +1,100 @@
+            } else if (val instanceof Map<?, ?>) {
+                map.put(key, convertAvroUtf8Map((Map<Object, Object>)val));
+            } else {
+                map.put(key, val);
```
Same comment about arrays/sets of Utf8 objects. We need an if/else for those as well.

vesense replied:
Updated.

vesense commented:
@sachin-goyal @sachingsachin Thanks for your review. I think I have addressed all of your comments.
Can you take a look again?

vesense commented:
@HeartSaVioR ready for merging after get responses from @sachin-goyal and @sachingsachin

sachingsachin commented on external/sql/storm-sql-runtime/src/jvm/org/apache/storm/sql/runtime/serde/avro/CachedSchemas.java:
```diff
@@ -0,0 +1,37 @@
+
+import java.util.WeakHashMap;
+
+public class CachedSchemas {
```
Please help me understand the use case for this class. As I see it, Only two classes use it and the `schemaString` they pass to `getSchema()` is final in those classes. So they will always pass the same string and get back the same object. Why not store the `Schema` object directly in those classes? If there is a good reason, we should add that comment here.

sachingsachin commented on external/sql/storm-sql-runtime/src/jvm/org/apache/storm/sql/runtime/serde/avro/AvroSerializer.java:
```diff
@@ -0,0 +1,72 @@
+/**
+ * AvroSerializer uses generic(without code generation) instead of specific(with code generation) writers.
+ */
+public class AvroSerializer implements IOutputSerializer, Serializable {
```
Would it be a good idea to merge AvroSerializer with AvroScheme? It seems they have the same state - schemaString, fieldNames and schemas. If only difference is the methods they have and the Interface they implement, then it may be simpler to merge the two classes into single AvroScheme?

vesense replied:
I prefer to not merge them. I think `inputFormatClass` and `outputFormatClass` are specifed by different classes that is more clearly to users.

sachingsachin commented on external/sql/storm-sql-runtime/src/jvm/org/apache/storm/sql/runtime/utils/SerdeUtils.java:
```diff
@@ -0,0 +1,115 @@
+        return serializer;
+    }
+
+    public static Object convertAvroUtf8(Object value){
```
+1 for this fully recursive function !

vesense replied:
Thanks.

vesense commented:
@sachingsachin Addressed.

sachingsachin commented:
@vesense - The pull request looks good. Thank you. Could you also add a small note on the JIRA STORM-2172 why you are trying to support avro? It would help to track it better in the future.

dossett commented:
storm-hdfs also has Avro (de)serializer functionality, including the ability to plug into different Avro schema registries (https://github.com/apache/storm/tree/master/external/storm-hdfs/src/main/java/org/apache/storm/hdfs/avro).  Can that functionality be refactored and shared in `storm-sql`?  Otherwise there will be two independent avro serialization schemes in Storm.

Apologies for the late comments on this change.

HeartSaVioR commented on external/sql/storm-sql-runtime/src/jvm/org/apache/storm/sql/runtime/utils/SerdeUtils.java:
```diff
@@ -0,0 +1,115 @@
+        Scheme scheme;
+        if (isNotEmpty(inputFormatClass)) {
+            switch (inputFormatClass) {
+                case "org.apache.storm.sql.runtime.serde.json.JsonScheme" :
```
You could avoid specifying full classname by importing class, and using JsonScheme.class.getName(). 
This also helps when need to shade: Please refer https://github.com/apache/calcite/pull/318

vesense replied:
Thanks @HeartSaVioR 
`case` clause required a constant expression, but I can change it to `if` clause to avoid hard coded.

HeartSaVioR commented on external/sql/storm-sql-runtime/src/jvm/org/apache/storm/sql/runtime/utils/SerdeUtils.java:
```diff
@@ -0,0 +1,115 @@
+                case "org.apache.storm.sql.runtime.serde.json.JsonScheme" :
+                    scheme = new JsonScheme(fieldNames);
+                    break;
+                case "org.apache.storm.sql.runtime.serde.avro.AvroScheme" :
```
Same on above.

HeartSaVioR commented on external/sql/storm-sql-runtime/src/jvm/org/apache/storm/sql/runtime/utils/SerdeUtils.java:
```diff
@@ -0,0 +1,115 @@
+        IOutputSerializer serializer;
+        if (isNotEmpty(outputFormatClass)) {
+            switch (outputFormatClass) {
+                case "org.apache.storm.sql.runtime.serde.json.JsonSerializer" :
```
Same on above.

HeartSaVioR commented on external/sql/storm-sql-runtime/src/jvm/org/apache/storm/sql/runtime/utils/SerdeUtils.java:
```diff
@@ -0,0 +1,115 @@
+                case "org.apache.storm.sql.runtime.serde.json.JsonSerializer" :
+                    serializer = new JsonSerializer(fieldNames);
+                    break;
+                case "org.apache.storm.sql.runtime.serde.avro.AvroSerializer" :
```
Same on above.

HeartSaVioR commented on external/sql/storm-sql-runtime/src/jvm/org/apache/storm/sql/runtime/utils/SerdeUtils.java:
```diff
@@ -0,0 +1,115 @@
+                    serializer = new JsonSerializer(fieldNames);
+                    break;
+                case "org.apache.storm.sql.runtime.serde.avro.AvroSerializer" :
+                    String schemaString = properties.getProperty("avro.schema");
```
Input schema and output schema can be different. We might need to pick other property name.

vesense replied:
Will fix.

vesense commented:
@sachingsachin 
>Could you also add a small note on the JIRA STORM-2172 why you are trying to support avro?

STORM-2172 is a part of STORM-2149, you can see the Jira Description for the details.

@dossett 
I agree with you for that something should be shared. Also, like @HeartSaVioR said, storm-sql and storm-hdfs have different interfaces, so we may have to spend time to rearrange it later.

@HeartSaVioR Addressed your comments.
The Avro Schema string is set through `Properties ` in DataSourcesProvider.constructTrident` which is specified by TBLPROPERTIES.

HeartSaVioR commented:
@vesense 
If this comes from TBLPROPERTIES, now TBLPROPERTIES is shared between data sources and input / output format. How we differentiate them? Would adding prefix be fine? Or would we want to manage reserved keys?

vesense commented:
@HeartSaVioR  You can take a look at the code after I addressed that using `input.avro.schema` and `output.avro.schema` for input and output Avro Schemas.

HeartSaVioR commented:
@vesense OK great. We could even document this so that other Storm SQL data source contributors don't use 'input' / 'output' as prefix.

HeartSaVioR commented:
@vesense Could you provide working example so that we can test it manually and finally document it? I'm +1 after checking manual test.

vesense commented:
example SQL:
_NOTE:  All the `\n` should be removed before you test it_
```
CREATE EXTERNAL TABLE ORDERS (ID INT PRIMARY KEY, UNIT_PRICE INT, QUANTITY INT)
 STORED AS
 INPUTFORMAT org.apache.storm.sql.runtime.serde.avro.AvroScheme
 OUTPUTFORMAT org.apache.storm.sql.runtime.serde.avro.AvroSerializer
 LOCATION 'kafka://localhost:2181/brokers?topic=orders'
 TBLPROPERTIES '
 {
	"producer": {
		"bootstrap.servers": "localhost:9092",
		"acks": "1",
		"key.serializer": "org.apache.org.apache.storm.kafka.IntSerializer",
		"value.serializer": "org.apache.org.apache.storm.kafka.ByteBufferSerializer"
	},
	
	"input.avro.schema": "
{\"type\": \"record\", 
 \"name\": \"orders\",
 \"fields\" : [
  {\"name\": \"id\", \"type\": \"int\"},
  {\"name\": \"unit_price\", \"type\": \"int\"},
  {\"name\": \"quantity\", \"type\": \"int\"},
  {\"name\": \"timestamp\", \"type\": \"long\"}
 ]
}
	",
	
	"output.avro.schema": "
{\"type\": \"record\", 
 \"name\": \"orders\",
 \"fields\" : [
  {\"name\": \"id\", \"type\": \"int\"},
  {\"name\": \"unit_price\", \"type\": \"int\"},
  {\"name\": \"quantity\", \"type\": \"int\"}
 ]
}
	"
}'

CREATE EXTERNAL TABLE LARGE_ORDERS (ID INT PRIMARY KEY, TOTAL INT)
 STORED AS
 INPUTFORMAT org.apache.storm.sql.runtime.serde.avro.AvroScheme
 OUTPUTFORMAT org.apache.storm.sql.runtime.serde.avro.AvroSerializer
 LOCATION 'kafka://localhost:2181/brokers?topic=large_orders'
 TBLPROPERTIES '
 {
	"producer": {
		"bootstrap.servers": "localhost:9092",
		"acks": "1",
		"key.serializer": "org.apache.org.apache.storm.kafka.IntSerializer",
		"value.serializer": "org.apache.org.apache.storm.kafka.ByteBufferSerializer"
	},
	
	"input.avro.schema": "
{\"type\": \"record\", 
 \"name\": \"large_orders\",
 \"fields\" : [
  {\"name\": \"id\", \"type\": \"int\"},
  {\"name\": \"total\", \"type\": \"int\"},
  {\"name\": \"timestamp\", \"type\": \"long\"}
 ]
}
	",
	
	"output.avro.schema": "
{\"type\": \"record\", 
 \"name\": \"large_orders\",
 \"fields\" : [
  {\"name\": \"id\", \"type\": \"int\"},
  {\"name\": \"total\", \"type\": \"int\"}
 ]
}
	"
}'

INSERT INTO LARGE_ORDERS
 SELECT ID, UNIT_PRICE * QUANTITY AS TOTAL FROM ORDERS WHERE UNIT_PRICE * QUANTITY > 50

```

vesense commented:
@HeartSaVioR I will create another PR for 1.x-branch because I find that `Utils.newInstance` has changed.
We need a type casting in 1.x-branch.

HeartSaVioR commented:
We might need to change the delimiter to ';', not '\n' since with Avro schema definition it becomes really hard to edit. (It was hard to edit indeed...)
Maybe need to initiate discussion around this from dev@. I'll handle this.

HeartSaVioR commented:
@vesense 
I got an exception while executing topology:

```
21:12:44.247 [main] INFO  o.a.s.s.r.DataSourcesRegistry - Registering scheme kafka with org.apache.storm.sql.kafka.KafkaDataSourcesProvider@338c99c8
Exception in thread "main" java.lang.IllegalStateException: Bolt 'b-0-LOGICALFILTER_6-LOGICALPROJECT_7' contains a non-serializable field of type org.apache.avro.Schema$RecordSchema, which was instantiated prior to topology creation. org.apache.avro.Schema$RecordSchema should be instantiated within the prepare method of 'b-0-LOGICALFILTER_6-LOGICALPROJECT_7 at the earliest.
	at org.apache.storm.topology.TopologyBuilder.createTopology(TopologyBuilder.java:127)
	at org.apache.storm.trident.topology.TridentTopologyBuilder.buildTopology(TridentTopologyBuilder.java:265)
	at org.apache.storm.trident.TridentTopology.build(TridentTopology.java:529)
	at org.apache.storm.sql.StormSqlImpl.submit(StormSqlImpl.java:134)
	at org.apache.storm.sql.StormSqlRunner.main(StormSqlRunner.java:63)
Caused by: java.lang.RuntimeException: java.io.NotSerializableException: org.apache.avro.Schema$RecordSchema
	at org.apache.storm.utils.Utils.javaSerialize(Utils.java:235)
	at org.apache.storm.topology.TopologyBuilder.createTopology(TopologyBuilder.java:122)
	... 4 more
Caused by: java.io.NotSerializableException: org.apache.avro.Schema$RecordSchema
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1184)
	at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1548)
	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1509)
	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1432)
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1178)
	at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1548)
	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1509)
...
```

Maybe `schema` shouldn't be parsed from constructor. (We could use it for verification but shouldn't store it.) 
Instead, you can initialize for the first time of write() or deserialize(), and reuse it after calling.

Please let me know if your patch works well with SQL runner with remote cluster.

HeartSaVioR commented:
Just an idea: We might even infer Avro schema from SQL table column information. The thing is associating between Avro type and SQL type. 
https://github.com/databricks/spark-avro seems to do similar.

HeartSaVioR commented:
Read JSON and calculate & filter, and store to AVRO:

```
CREATE EXTERNAL TABLE ORDERS (ID INT PRIMARY KEY, UNIT_PRICE INT, QUANTITY INT) STORED AS INPUTFORMAT 'org.apache.storm.sql.runtime.serde.json.JsonScheme' OUTPUTFORMAT 'org.apache.storm.sql.runtime.serde.json.JsonSerializer' LOCATION 'kafka://localhost:2181/brokers?topic=orders' TBLPROPERTIES '{ "producer": { "bootstrap.servers": "localhost:9092", "acks": "1", "key.serializer": "org.apache.storm.kafka.IntSerializer", "value.serializer": "org.apache.storm.kafka.ByteBufferSerializer" } }'
CREATE EXTERNAL TABLE LARGE_ORDERS (ID INT PRIMARY KEY, TOTAL INT) STORED AS INPUTFORMAT 'org.apache.storm.sql.runtime.serde.avro.AvroScheme' OUTPUTFORMAT 'org.apache.storm.sql.runtime.serde.avro.AvroSerializer' LOCATION 'kafka://localhost:2181/brokers?topic=large_orders' TBLPROPERTIES ' { "producer": { "bootstrap.servers": "localhost:9092", "acks": "1", "key.serializer": "org.apache.storm.kafka.IntSerializer", "value.serializer": "org.apache.storm.kafka.ByteBufferSerializer" }, "input.avro.schema": "{\"type\": \"record\", \"name\": \"large_orders\", \"fields\" : [ {\"name\": \"ID\", \"type\": \"int\"}, {\"name\": \"TOTAL\", \"type\": \"int\"} ]} ", "output.avro.schema": "{\"type\": \"record\", \"name\": \"large_orders\", \"fields\" : [ {\"name\": \"ID\", \"type\": \"int\"}, {\"name\": \"TOTAL\", \"type\": \"int\"} ]} "}'
```

Read AVRO and store to JSON:

```
CREATE EXTERNAL TABLE LARGE_ORDERS (ID INT PRIMARY KEY, TOTAL INT) STORED AS INPUTFORMAT 'org.apache.storm.sql.runtime.serde.avro.AvroScheme' OUTPUTFORMAT 'org.apache.storm.sql.runtime.serde.avro.AvroSerializer' LOCATION 'kafka://localhost:2181/brokers?topic=large_orders' TBLPROPERTIES ' { "producer": { "bootstrap.servers": "localhost:9092", "acks": "1", "key.serializer": "org.apache.storm.kafka.IntSerializer", "value.serializer": "org.apache.storm.kafka.ByteBufferSerializer" }, "input.avro.schema": "{\"type\": \"record\", \"name\": \"large_orders\", \"fields\" : [ {\"name\": \"ID\", \"type\": \"int\"}, {\"name\": \"TOTAL\", \"type\": \"int\"} ]} ", "output.avro.schema": "{\"type\": \"record\", \"name\": \"large_orders\", \"fields\" : [ {\"name\": \"ID\", \"type\": \"int\"}, {\"name\": \"TOTAL\", \"type\": \"int\"} ]} "}'
CREATE EXTERNAL TABLE LARGE_ORDERS_JSON (ID INT PRIMARY KEY, TOTAL INT) STORED AS INPUTFORMAT 'org.apache.storm.sql.runtime.serde.json.JsonScheme' OUTPUTFORMAT 'org.apache.storm.sql.runtime.serde.json.JsonSerializer' LOCATION 'kafka://localhost:2181/brokers?topic=large_orders_json' TBLPROPERTIES '{ "producer": { "bootstrap.servers": "localhost:9092", "acks": "1", "key.serializer": "org.apache.storm.kafka.IntSerializer", "value.serializer": "org.apache.storm.kafka.ByteBufferSerializer" } }'
INSERT INTO LARGE_ORDERS_JSON SELECT ID, TOTAL FROM LARGE_ORDERS
```

Manual test succeed. +1 
Thanks for the great work @vesense

vesense commented:
Yes, I also find that the sql with Avro schema string is a little terrible. it's not easy to edit for users. maybe I can take time to improve later.

vesense commented:
@HeartSaVioR Thanks for your patience.
After this PR getting merged, I will modify the PR for TSV/CSV format ASAP.

HeartSaVioR commented:
@vesense Could you craft a patch for 1.x branch? It should be compiled with JDK 1.7.

This is build failure with JDK 1.7 on 1.x-branch & this patch.

```
[INFO] -------------------------------------------------------------
[ERROR] COMPILATION ERROR :
[INFO] -------------------------------------------------------------
[ERROR] /Users/jlim/WorkArea/JavaProjects/storm/external/sql/storm-sql-runtime/src/test/org/apache/storm/sql/TestAvroSerializer.java:[47,40] incompatible types
  required: java.util.List<java.lang.Object>
  found:    java.util.ArrayList<java.lang.Object&java.io.Serializable&java.lang.Comparable<? extends java.lang.Object&java.io.Serializable&java.lang.Comparable<?>>>
[INFO] 1 error
[INFO] -------------------------------------------------------------
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO]
[INFO] storm-sql-runtime .................................. FAILURE [  2.960 s]
[INFO] storm-sql-core ..................................... SKIPPED
[INFO] storm-sql-kafka .................................... SKIPPED
[INFO] storm-sql-redis .................................... SKIPPED
[INFO] storm-sql-mongodb .................................. SKIPPED
[INFO] sql ................................................ SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 3.844 s
[INFO] Finished at: 2016-11-11T23:11:46+09:00
[INFO] Final Memory: 35M/452M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.1:testCompile (default-testCompile) on project storm-sql-runtime: Compilation failure
[ERROR] /Users/jlim/WorkArea/JavaProjects/storm/external/sql/storm-sql-runtime/src/test/org/apache/storm/sql/TestAvroSerializer.java:[47,40] incompatible types
[ERROR] required: java.util.List<java.lang.Object>
[ERROR] found:    java.util.ArrayList<java.lang.Object&java.io.Serializable&java.lang.Comparable<? extends java.lang.Object&java.io.Serializable&java.lang.Comparable<?>>>
[ERROR] -> [Help 1]
```

vesense commented:
@HeartSaVioR OK. I will create a PR for 1.x-branch
